Initial meeting tag support

pokey opened the pull request:
- Partially addresses https://github.com/knausj85/knausj_talon/issues/835
- Blocked by https://github.com/talonvoice/talon/issues/606

pokey committed:
Initial meeting tag support

pokey commented on apps/zoom/zoom.py:
```diff
@@ -0,0 +1,45 @@
+@ctx.action_class("user")
+class UserActions:
+    def meeting_mute():
+        app.notify("Mute meeting")
+
+    def meeting_unmute():
+        app.notify("Unmute meeting")
+
+    def meeting_exit():
+        app.notify("Exit meeting")
```
@nriley here's where you'd put the implementations of mute / unmute / exit.  Need to decide whether we're comfortable using accessibility here.  Also prob need to make the tag mac-specific if we use accessibility

pokey commented on apps/zoom/zoom.py:
```diff
@@ -0,0 +1,45 @@
+
+
+def is_zoom(app):
+    return app.bundle == "us.zoom.xos"
```
Todo: make this work on Windows

pokey commented on apps/zoom/zoom.talon:
Maybe we should rename this file?  Because this file name should really be for when Zoom is focused

pokey commented:
@nriley feel free to push directly to this branch, or open a PR against it if you can't for whatever reason

nriley committed:
Initial mute/unmute/exit support for Zoom on Mac.

Also adds user.meeting_is_muted.

Tested in US English, German and Simplified Chinese localizations.

nriley commented:
Instead of looking at the menu (with a different title per-language) I can recognize the shape of the "you are muted" icon and determine the mute status this way; it worked across the localizations and meeting window layouts I tried. Still might be some issues (even ignoring the Talon bug) in recognizing the meeting window title.

If we just wanted to support toggling mute, we could do it without accessibility.

I put in a partial workaround for the bug by focusing another app and back to Zoom. I don't think actually checking the 2nd time should help in most cases but it should help Talon to notice the window for **next** time. The problem is still though that the tag won't activate, so the voice commands don't do anything (other than "media mute" as @pokey experienced yesterday :).

One thing to note while I was testing, which in part inspired my two-microphone solution: it seems that if you mute audio in Zoom, you can't talk to Talon either **on the same microphone** (works fine with another microphone).

Another action that might be helpful would be to allow you to focus the active meeting window since I've certainly had cases where it gets "lost". This wouldn't require accessibility either (at least for Zoom), since my window filtering just uses the title.

nriley committed:
Merge branch 'main' into pokey/initial-meeting-tag-support

nriley commented:
Here's a thought of a slightly more robust workaround for the problem of meeting commands not being active since Talon hasn't "found" the meeting window yet.

We could have the meeting commands active if *any* of the meeting apps are active, using "user.running". Then in the default implementation for "meeting mute", etc., we could "poke" each of the apps (via accessibility to find the window) to see if we're failing to notice an active meeting window.

Another option is to poll windows in active meeting apps by accessibility (which I may end up doing in my own repo, so I can accurately switch microphones) and "force" Talon to notice any it hasn't found. I do that in one of my private repos, but I'd be a bit worried about doing that by default.

nriley committed:
Initial mute/unmute/exit support for Teams on Mac.

Could not figure out any way to determine mute status that isn't language-specific or talon.experimental.locate, so just notify if we can't figure it out. "Meeting mute" will toggle mute/unmute regardless.

nriley commented:
Added Teams support — due to its use of Electron this is more complex/asynchronous. Given Teams is pending a rewrite I'm disinclined to spend much more time on this particular implementation, but it at least works for me. Talon seems to be more reliably "see" the meeting window in Teams, so I didn't include the same workaround I did for Zoom.

nriley committed:
Use actions when a meeting starts/stops.

This allows you to react to starting/stopping a meeting by overriding these actions, and using actions.next to invoke the default implementation.

Also, remove redundant check for a meeting window in Teams and Zoom on Talon launch.

nriley commented:
Refactored a bit to put the `user.meeting_appname` tag context in one place rather than creating a context per meeting app. At this point we could also just set user.meeting in the module implementation rather than using .talon files to do so, but I left that part alone for now as I don't feel strongly one way or the other.

Here's where I override the `user.meeting_started`/`user.meeting_ended` in my `suspend` script to turn off or switch the microphone during a meeting: https://github.com/nriley/knausj_talon/commit/26ac780f5cc0191de97f7eb3c648c01d07f7643a

nriley committed:
Remove overly specific check for Teams window title.

nriley committed:
Don't send redundant "meeting ended" notifications for Teams.

nriley committed:
Don't toggle the meeting tag on a redundant meeting_ended action.

nriley committed:
Fix incorrect variable when detecting an active Zoom meeting on Talon launch or module reload.

nriley committed:
Initial mute/unmute/exit support for Webex Meetings on Mac.

nriley committed:
Remove unused import.

auscompgeek commented:
> I can recognize the shape of the "you are muted" icon and determine the mute status this way

Does this work even when the Zoom meeting controls are set to auto-hide?

nriley commented:
> > I can recognize the shape of the "you are muted" icon and determine the mute status this way
> 
> Does this work even when the Zoom meeting controls are set to auto-hide?

Yes, I think so — the icon I am matching shows up even if the controls are hidden.

<img width="110" alt="image" src="https://user-images.githubusercontent.com/68303/234038205-99e45b70-aefe-413e-83c8-d60f83e2c092.png">

If you can find cases where the above mute icon does not appear, then I can implement checking the menus in addition (like I do on Webex) — unfortunately this will not be robust to non-English UI languages.

nriley committed:
Stay in a meeting when you are screen sharing with Webex Meetings on Mac.

nriley committed:
Match menu item to determine if Zoom is muted.

Unfortunately this will break in UI languages other than English.

nriley commented:
OK, I've moved off my slick-but-broken Zoom mute detection.

More testing appreciated!

auscompgeek commented:
Does the Zoom menu bar item need to be enabled for that to work, or does it use the menus for the meeting window?

nriley commented:
Yes, it does not use the menu in the top right corner of the screen but instead uses the application menus even if the application is not active.

nriley committed:
Resolve race between declaration and use of user.meeting_started action.

nriley committed:
Use an alternative DOM identifier to recognize when we are in a Teams meeting.

Also better explain why each identifier is used - the same window is shared with "pre-joining", but if you are switching between the smaller and larger versions of the meeting window, the first identifier exists in both.

(Suspect the prior identifier is no longer being used as of Teams 1.6.00.11156.)

nriley commented:
Since I upgraded either to a recent Talon beta or macOS (or both), this has become incredibly unreliable. (It's blocked on the window activation callback issue in core Talon, but it was at least usable previously.) Will work on fixing it as I have time.

nriley commented:
https://github.com/talonvoice/talon/issues/606 is fixed in the latest beta! I tested Zoom and it seems reliable.
